### PR TITLE
feat: donate page update

### LINKED ISSUE #37617: freecodecamp.org/donate should explain it is a donate page, even for unauthenticated users
Currently, when an unauthenticated user visits https://freecodecamp.org/donate, it shows a sign-in page with no indication it is a donate page.

It would be a better user experience if there was a message at the top of the page saying it was a donate page and explaining why authentication is necessary. With no message about this, users could assume that they are at the wrong page to donate or there is an error happening.

### PR DESCRIPTION
**Pull request by raisedadead**
This pull-request addresses several new updates to our donation infrastructure:

- [x] two column layout for the page.
- [x] amount to learning minutes mapping for contribution impact.
- [x] handle one-time and recurring stripe subscription charges.
- [x] server side validation of donate forms.
- [x] prevent multiple subscriptions and onetime donations per user.

Via Servicebot
- [x] change stripe subscription plan?
- [x] cancel a subscription plan programatically?
- [x] download invoices from Stripe.
- [x] update card on Stripe.

WIP:
- [ ] PayPal integration for subscription and onetime donations.

Planned:
- [ ] normalize non-donors (canceled subscriptions) in the database?
- [ ] Send a plan id instead of amount and duration to the API.

### CODE REVIEW

**Review by raisedadead — COMMENTED** [review]
Todo (must for acceptance) :
- [x] Unify donation config on server and client.
- [x] Validate if isDonating already before creating a subscription.

**Comment by ahmadabdolsaheb** [comment]
should we change the text of "one-time" with single for mobile?

image

**Comment by raisedadead** [comment]
> should we change the text of "one-time" with single for mobile?

We should prevent the wrap?

**Comment by raisedadead** [comment]
I have opened this for review, with the thought that the remaining WIP and planned items will come in a separate PRs including comprehensive tests.

File: client/src/pages/donation/settings.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        <DonateServicebotEmbed email={currentSettingsEmail} hash={hash} />
+      </div>
+    );
+  }
+
+  renderDonationEmailsList() {
+    const { donationEmails } = this.props;
+    return (
+      <div>
+        {donationEmails.map((email, index) => (
```
Comment (raisedadead) [inline client/src/pages/donation/settings.js]: `donationEmails` array may have duplicates. Should do a cleanup with lodash or something before iterating and creating the buttons.

File: client/src/components/Donation/components/DonateForm.js [hunk 2]
```diff
@@ -44,66 +64,197 @@ const createOnClick = navigate => e => {
...
+                </ToggleButtonGroup>
+                <Spacer />
+                <p className='donation-description'>
+                  {`Your `}
+                  {this.getFormatedAmountLabel(donationAmount)}
+                  {` donation will provide `}
+                  {this.convertToTimeContributed(donationAmount)}
+                  {` of learning to people around the world `}
+                  {duration === 'onetime' ? `for one ` : `each `}
+                  {duration === 'month' ? `month.` : `year.`}
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/components/DonateForm.js]: For onetime donations I think it reads better in the following form:

> Your $35 donation will provide 1,750 hours of learning to people around the world.

File: client/src/components/layouts/Default.js [hunk 3]
```diff
@@ -145,13 +146,16 @@ class DefaultLayout extends Component {
...
-      theme = 'default'
+      theme = 'default',
+      useTheme = true
     } = this.props;
     return (
       <Fragment>
         <Helmet
           bodyAttributes={{
-            class: `${theme === 'default' ? 'light-palette' : 'dark-palette'}`
+            class: useTheme
```
Comment (ojeytonwilliams) [inline client/src/components/layouts/Default.js]: Is this just intended to be until a dark theme version is created?

File: client/src/pages/donate.js [hunk 4]
```diff
@@ -72,21 +85,44 @@ export class DonatePage extends Component {
...
+                stripe={stripe}
+              />
+              <Row>
+                <Col sm={10} smOffset={1} xs={12}>
+                  <Spacer size={2} />
+                  <h3 className='text-center'>Manage your existing donation</h3>
+                  {[
+                    `Update your existing donation`,
+                    `Download donation receipts`
+                  ].map(donationSettingOps => (
```
Comment (ojeytonwilliams) [inline client/src/pages/donate.js]: It looks like these two buttons both just point to the same place. Should they do different things?

File: client/src/pages/donation/settings.js [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+  handleSelectDonationEmail(e) {
+    e.preventDefault();
+    this.setState({ hash: null, currentSettingsEmail: null });
+    this.getEmailHmacHash(e.currentTarget.value);
+  }
+
+  renderServicebotEmbed() {
+    const { currentSettingsEmail, hash } = this.state;
+    if (!hash && !currentSettingsEmail) {
```
Comment (ojeytonwilliams) [inline client/src/pages/donation/settings.js]:
```suggestion
    if (!hash || !currentSettingsEmail) {
```
It needs both, no?

File: client/src/components/Donation/components/DonateServicebotEmbed.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      servicebot_id: servicebotId,
+      service: 'freeCodeCamp.org',
+      email,
+      hash,
+      options: {
+        cancel_now: true,
+        disableCoupon: true,
+        forceCard: true,
+        disableTiers: [
+          'Monthly $10 Donation  - Unavailable',
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/components/DonateServicebotEmbed.js]:
```suggestion
          'Monthly $10 Donation - Unavailable',
```

File: client/src/pages/donation/settings.js [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        <DonateServicebotEmbed email={currentSettingsEmail} hash={hash} />
+      </div>
+    );
+  }
+
+  renderDonationEmailsList() {
+    const { donationEmails } = this.props;
+    return (
+      <div>
+        {uniq(donationEmails).map((email, index) => (
```
Comment (ojeytonwilliams) [inline client/src/pages/donation/settings.js]:
```suggestion
        {uniq(donationEmails).map(email => (
```
Emails are already unique keys

File: client/src/pages/donation/settings.js [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+          <div key={email + '-' + index}>
```
Comment (ojeytonwilliams) [inline client/src/pages/donation/settings.js]:
```suggestion
          <div key={email}>
```

File: client/src/pages/donation/settings.js [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  }
+
+  render() {
+    const { showLoading, isSignedIn, isDonating, navigate } = this.props;
+
+    if (showLoading) {
+      return <Loader fullScreen={true} />;
+    }
+
+    if (!showLoading && !isSignedIn) {
```
Comment (ojeytonwilliams) [inline client/src/pages/donation/settings.js]:
```suggestion
    if (!isSignedIn) {
```
If we're here, `showLoading` is false.

File: client/src/pages/donation/settings.js [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+      navigate(`${apiLocation}/signin?returnTo=donation/settings`);
+      return <Loader fullScreen={true} />;
+    }
+
+    if (!showLoading && !isDonating) {
```
Comment (ojeytonwilliams) [inline client/src/pages/donation/settings.js]:
```suggestion
    if (!isDonating) {
```
Same logic as above.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
For the most part, this looks good!  I have a few nitpicks and queries:

File: api-server/server/boot/donate.js [hunk 11]
```diff
@@ -1,160 +1,269 @@
...
     });
   }
 
   function createStripePlan(plan) {
+    log(`Creating subscription plan: ${plan.product.name}`);
     stripe.plans.create(plan, function(err) {
       if (err) {
-        console.log(err);
-        throw err;
+        log(err);
```
[the remark this answers is not in the document]
Comment (raisedadead) [reply api-server/server/boot/donate.js]: No, we are not going to be creating plans programatically that often (it more like a one-off). Throwing errors will kill the server on all errors. Logging should be fine.

File: client/src/components/layouts/Default.js (hunk 3 above)
Comment (raisedadead) [reply client/src/components/layouts/Default.js]: Yes, the current servicebot css gets messed up with our darkmode styles.

**Comment by ojeytonwilliams** [comment]
@raisedadead if you want to merge this today, feel free to dismiss my review (I didn't find anything crucial, after all).  Otherwise I'll take another look tomorrow.

File: client/src/pages/donate.js (hunk 4 above)
Comment (raisedadead) [reply client/src/pages/donate.js]: This is an intended behaviour, we wanted to let users know at the first glance that they can do different things.

**Comment by raisedadead** [comment]
Thanks for those changes, I have pulled in those suggestions into the PR. I should be able to write tests soon too.

**Pull request merged**